firmware/lm32/heartbeat.c

```diff
@@ -0,0 +1,83 @@
+
+#include <generated/csr.h>
+#include <generated/mem.h>
+#include <hw/flags.h>
+#include <system.h>
+#include <time.h>
+
+#include "heartbeat.h"
+#include "processor.h"
+#include "hdmi_in0.h"
+#include "hdmi_in1.h"
+#include "pattern.h"
+
+static bool heartbeat_status = 0;
+
+void hb_status(bool val)
+{
+	if(val==1) { 
+		heartbeat_status = 1; 
+	}
+	else {
+		heartbeat_status = 0;
+	}		
+}
+
+void hb_service(int sink)
+{
+	static int last_event;
+	static int counter;		
+	static bool color_v;
+	if (heartbeat_status==1) {	
+		if(elapsed(&last_event, identifier_frequency_read()/120)) {
+			counter = counter+1;
+			hb_fill(color_v, sink);
+			if(counter>60){
+				color_v = !color_v;
+				counter = 0;
+			}
+		}
+	}
+}
+
+void hb_fill(bool color_v, int sink)
+{
+	int addr, i, j;
+	unsigned int color;
+
+	volatile unsigned int *framebuffer = (unsigned int *)(MAIN_RAM_BASE + pattern_framebuffer_base());
```

Comment:
`volatile` not explicitly required here, but will be continued to use for consistency. `volatile` is used when we want our memory accesses for that variable to be ordered.